source/shared/ed25519-lc.c

```diff
@@ -0,0 +1,379 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/cal/ed25519.h>
+#include <aws/cal/private/opensslcrypto_common.h>
+
+#include <aws/common/device_random.h>
+#include <aws/common/encoding.h>
+#include <aws/common/ref_count.h>
+
+#include <openssl/evp.h>
+
+struct aws_ed25519_key_pair {
+    struct aws_allocator *allocator;
+    struct aws_ref_count ref_count;
+
+    EVP_PKEY *key;
+};
+
+static void s_ed25519_destroy_key(void *key_pair) {
+    if (key_pair == NULL) {
+        return;
+    }
+
+    struct aws_ed25519_key_pair *lc_key_pair = (struct aws_ed25519_key_pair *)(key_pair);
+
+    if (lc_key_pair->key != NULL) {
+        EVP_PKEY_free(lc_key_pair->key);
+    }
+
+    aws_mem_release(lc_key_pair->allocator, lc_key_pair);
+}
+
+struct aws_ed25519_key_pair *aws_ed25519_key_pair_new_generate(struct aws_allocator *allocator) {
+#if defined(OPENSSL_IS_OPENSSL) && OPENSSL_VERSION_NUMBER <= 0x10101000L
+    /* ed25519 support does not exist prior to 1.1.1 */
+    aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    return NULL;
+#else
+    EVP_PKEY *pkey = NULL;
+
+    AWS_LOGF_DEBUG(0, "here1");
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, NULL);
+    if (ctx == NULL) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        return NULL;
```

Comment:
maybe
`aws_reinterpret_lc_evp_error_as_crt(0//error, "EVP_PKEY_CTX_new_id", AWS_LS_CAL_ED25519)`?

Reply:
EVP_PKEY_CTX_new_id does not set any error params as far as i can tell, so running it through reinterp is not useful